pkg/importer/dataStream.go

```diff
@@ -970,49 +895,36 @@ func (d *DataStream) convertQcow2ToRaw(src, dest string) error {
 }
 
 // Copy endpoint to dest based on passed-in reader.
-func (d *DataStream) copy(dest string) (err error) {
+func (d *DataStream) copy(dest string) error {
+	if util.GetAvailableSpace(d.ScratchDataDir) > int64(0) {
+		defer CleanDir(d.ScratchDataDir)
+	}
 	if d.isHTTPQcow2() {
-		err = d.convertQcow2ToRawStream(dest)
+		err := d.convertQcow2ToRawStream(dest)
 		if err != nil {
 			return err
 		}
 	} else {
-		dest = filepath.Clean(dest)
-		klog.V(2).Infof("copying image file to %q", dest)
-		tmpDest := dest
-		if !d.isQemuStored() {
-			//it is either not qemu or not stored
-			if d.qemu {
-				// copy to tmp; qemu conversion will write to passed-in destination
-				tmpDest = randTmpName(dest)
-				klog.V(3).Infof("Copy: temp file for qcow2 conversion: %q", tmpDest)
-				defer func(f string) {
-					os.Remove(f)
-				}(tmpDest)
-			}
-			if d.isMoveable() {
-				klog.V(3).Infof("Moving already stored raw from %s to %s", d.tmpDataPath, tmpDest)
-				err = MoveFile(d.tmpDataPath, tmpDest)
-				if err != nil {
-					klog.V(3).Infof("Failing back to copying stored raw from %s to %s", d.tmpDataPath, tmpDest)
-					//Failback to copy since move failed
-					err = StreamDataToFile(d.topReader(), tmpDest)
-				}
-			} else {
-				klog.V(3).Infof("Streaming data of non Movble file to %s", tmpDest)
-				err = StreamDataToFile(d.topReader(), tmpDest)
-			}
-			if err != nil {
-				return errors.WithMessage(err, fmt.Sprintf("unable to stream data to file %q", dest))
-			}
-		} else {
-			//It is a stored qemu file - taking a location where it is stored
-			klog.V(3).Infof("Streaming file to %s", tmpDest)
-			tmpDest = d.tmpDataPath
+		if util.GetAvailableSpace(d.ScratchDataDir) <= int64(0) {
```

Comment:
@awels  may be during refactoring it is worth considering to place the logic that decides whether scratch space is required or not with import method logic, otherwise we have  several places that check and verify weather scratch space is required and is provided, which is confusing IMO.

For example it is enough to decide per import method that Scratch is required and then use it inside that method (http/s3/registry etc.) Scratch can be assigned to dataStreamer (http/s3/registry) and in code that is common to all streamers (as copy) the scratch can be taken from concrete streamer if it is required. In such case all the logic is in one place per importing technique (see https://github.com/kubevirt/containerized-data-importer/pull/633 to see what I mean)